README.rst
CGHub Python API
================

CGHub Python API.

Installation
------------

To get the latest commit from GitHub

.. code-block:: bash

    $ pip install -e git+git://github.com/42cc/cghub-python-api.git#egg=cghub_python_api


Usage
-----

.. code-block:: python

    >>> from cghub_python_api import WSAPIRequest as Request
    >>> results = []
    >>> request = Request(query={'analysis_id': ['016cc2ca-db11-44af-b846-18e9275a4c4b', '01a589db-02d8-4d75-a2da-bb0bd8140a32']})
    >>> for result in request.call():
    ...     print result.state.text
    ...     results.append(result)
    ...
    live
    live
    >>> print results[0].disease_abbr.text
    BRCA
    >>> print results[0]['disease_abbr'].text
    BRCA
    >>> print results[1]['filename'].text
    UNCID_1176640.2bdc311e-59cc-449f-b8dc-6662052678fd.sorted_genome_alignments.bam
    >>> print results[1]['filename.0'].text
    UNCID_1176640.2bdc311e-59cc-449f-b8dc-6662052678fd.sorted_genome_alignments.bam
    >>> print results[1]['filename.1'].text
    UNCID_1176640.2bdc311e-59cc-449f-b8dc-6662052678fd.sorted_genome_alignments.bam.bai
    >>> print results[1]['checksum.0'].attrib
    {'type': 'MD5'}
    >>> results[0].analysis_id.exist
    True
    >>> results[0].somebadattr
    <cghub_python_api.api.NonExistent object at 0x1ef1950>
    >>> results[0].somebadattr.exist
    False
    >>> print results[0].somebadattr.text
    None


Read more in ``docs/source/*.rst``.


Running tests
-------------

.. code-block:: bash

    python -m unittest -v tests


Contribute
----------

If you want to contribute to this project, please perform the following steps

.. code-block:: bash

    # Fork this repository
    # Clone your fork
    $ cd cghub-python-api
    $ virtualenv .env --no-site-packages
    $ source .env/bin/activate
    $ python setup.py install
    $ pip install -r test_requirements.txt

    $ git co -b feature_branch master
    # Implement your feature and tests
    $ git add . && git commit
    $ git push -u origin feature_branch
    # Send us a pull request for your feature branch


CHANGELOG.txt
=== 0.1 ===
- Initial commit

=== 0.2 ===
- Ability to work with SOLR

=== 0.21 ===
- API accepts wsapi/solr json response


tests/__init__.py
from .test_utils import *
from .test_request import *


tests/test_utils.py
import os
import unittest
import urllib2

from mock import patch
from StringIO import StringIO

from cghub_python_api.utils import urlopen


class UtilsTestCase(unittest.TestCase):

    TEST_URL = 'https://192.35.223.223/cghub/metadata/analysisDetail'
    TEST_QUERY = {
            'last_modified': '[NOW-64DAY+TO+NOW-63DAY]',
            'state': 'live'}
    TEST_FILE_PATH = os.path.join(
                            os.path.dirname(__file__),
                            'data/wsapi_details_10_results.xml')

    def test_urlopen(self):
        f = open(self.TEST_FILE_PATH, 'r')
        data = f.read()
        io = StringIO(data)
        with patch.object(urllib2, 'urlopen', return_value=io) as urllib2_urlopen_mock:
            url = '%s?%s' % (self.TEST_URL, '&'.join(
                    ['%s=%s' for key, value in self.TEST_QUERY.iteritems()]))
            f = urlopen(url=url)
            data = f.read()
            self.assertIn('<analysis_id>', data)
            self.assertEqual(urllib2_urlopen_mock.call_count, 1)

    def test_urlopen_repeat_requests_on_urlerror(self):
        def raise_urlerror(req):
            raise urllib2.URLError('URL Error message')
        with patch('urllib2.urlopen') as urllib2_urlopen_mock:
            urllib2_urlopen_mock.side_effect = raise_urlerror
            url = '%s?%s' % (self.TEST_URL, '&'.join(
                    ['%s=%s' for key, value in self.TEST_QUERY.iteritems()]))
            attempts_count = 3
            try:
                f = urlopen(url=url, max_attempts=attempts_count)
            except urllib2.URLError as e:
                self.assertIn('%d attempts' % attempts_count, str(e))
            self.assertEqual(urllib2_urlopen_mock.call_count, attempts_count)

    def test_error_message_on_bad_request(self):
        message = 'Bad request'
        def raise_httperror(req):
            raise urllib2.HTTPError(
                    url='http://example.com', code=400,
                    msg=message, hdrs=None, fp=None)
        with patch('urllib2.urlopen') as urllib2_urlopen_mock:
            urllib2_urlopen_mock.side_effect = raise_httperror
            url = '%s?%s' % (self.TEST_URL, '&'.join(
                    ['%s=%s' for key, value in self.TEST_QUERY.iteritems()]))
            attempts_count = 3
            try:
                f = urlopen(url=url, max_attempts=attempts_count)
            except urllib2.URLError as e:
                self.assertNotIn('%d attempts' % attempts_count, str(e))
                self.assertIn(message, str(e))
            self.assertEqual(urllib2_urlopen_mock.call_count, 1)


if __name__ == '__main__':
    unittest.main()


tests/test_request.py
import os.path
import unittest
import urllib2

from StringIO import StringIO
from mock import patch

from cghub_python_api import WSAPIRequest, SOLRRequest


class RequestTestCase(unittest.TestCase):

    REQUEST_CLASS = WSAPIRequest
    TEST_FILE_PATH = os.path.join(
                            os.path.dirname(__file__),
                            'data/wsapi_details_10_results.xml')
    TEST_JSON_FILE_PATH = os.path.join(
                            os.path.dirname(__file__),
                            'data/wsapi_details_1_result.json')
    TEST_QUERY = {
                'last_modified': '[NOW-64DAY+TO+NOW-63DAY]',
                'state': 'live'}
    QUERY_TEST_DATA_SET = [
        {
            'query': {
                'study': ['phs000178', '*Other_Sequencing_Multiisolate', 'phs0004*'],
                'disease_abbr': ['BLCA', 'OR', 'BRCA'],
                'state': 'live'},
            'kwargs': {},
            'url': 'https://192.35.223.223/cghub/metadata/analysisDetail?'
                'study=(phs000178+OR+%2AOther_Sequencing_Multiisolate+OR+phs0004%2A)&'
                'disease_abbr=(BLCA+OR+OR+OR+BRCA)&state=live'
        }, {
            'query': {
                'study': '*Other_Sequencing_Multiisolate',
                'state': ['live'],
                'last_modified': '[NOW-1MONTH+TO+NOW]'},
            'kwargs': {},
            'url': 'https://192.35.223.223/cghub/metadata/analysisDetail?'
                'study=%2AOther_Sequencing_Multiisolate&'
                'last_modified=%5BNOW-1MONTH%2BTO%2BNOW%5D&state=(live)'
        }, {
            'query': {
                'study': '*Other_Sequencing_Multiisolate',
                'state': ['live']},
            'kwargs': {'limit': 10, 'offset': 10, 'sort_by': '-state'},
            'url': 'https://192.35.223.223/cghub/metadata/analysisDetail?'
                'study=%2AOther_Sequencing_Multiisolate&state=(live)&'
                'start=10&rows=10&sort_by=state:desc'
        }
    ]

    def _test_access_to_attributes(self, result):
        self.assertEqual(
                result.analysis_id.text,
                '3b4a2267-fafc-4185-ad44-bf0daadca912')
        self.assertEqual(
                result['analysis_id'].text,
                '3b4a2267-fafc-4185-ad44-bf0daadca912')
        self.assertEqual(result.analysis_id.exist, True)
        self.assertEqual(result['analysis_id'].exist, True)
        self.assertEqual(result.badattr.exist, False)
        self.assertEqual(result['badattr'].exist, False)
        self.assertEqual(result.badattr.text, None)
        file_attr = result['filename.0']
        self.assertTrue(file_attr.exist)
        self.assertTrue(file_attr.text.endswith('.bam'))
        file_attr = result['filename.1']
        self.assertTrue(file_attr.exist)
        self.assertTrue(file_attr.text.endswith('.bai'))

    def test_request(self):
        f = open(self.TEST_FILE_PATH, 'r')
        with patch.object(urllib2, 'urlopen', return_value=f) as urllib2_urlopen_mock:
            request = self.REQUEST_CLASS(query=self.TEST_QUERY)
            results = []
            for result in request.call():
                results.append(result)
            self.assertEqual(len(results), 10)
            self.assertEqual(request.hits, 10)
            self._test_access_to_attributes(results[0])
            self.assertEqual(urllib2_urlopen_mock.call_count, 1)

    def test_json_request(self):
        f = open(self.TEST_JSON_FILE_PATH, 'r')
        with patch.object(urllib2, 'urlopen', return_value=f) as urllib2_urlopen_mock:
            request = self.REQUEST_CLASS(
                    query=self.TEST_QUERY,
                    format=self.REQUEST_CLASS.FORMAT_JSON)
            results = []
            for result in request.call():
                results.append(result)
            self.assertEqual(len(results), 1)
            self.assertTrue(results[0]['analysis_id'])
            self.assertEqual(urllib2_urlopen_mock.call_count, 1)

    def test_request_build_query(self):
        f = open(self.TEST_FILE_PATH, 'r')
        data = f.read()
        io = StringIO(data)
        for data in self.QUERY_TEST_DATA_SET:
            io.seek(0)
            with patch.object(self.REQUEST_CLASS, 'get_source_file', return_value=io) as get_source_file_mock:
                request = self.REQUEST_CLASS(query=data['query'], **data['kwargs'])
                for i in request.call():
                    pass
                get_source_file_mock.assert_called_once_with(data['url'])


class SOLRRequestTestCase(RequestTestCase):

    REQUEST_CLASS = SOLRRequest
    TEST_FILE_PATH = os.path.join(
                            os.path.dirname(__file__),
                            'data/solr_details_10_results.xml')
    TEST_JSON_FILE_PATH = os.path.join(
                            os.path.dirname(__file__),
                            'data/solr_details_1_result.json')
    QUERY_TEST_DATA_SET = [
        {
            'query': {
                'study': ['phs000178', '*Other_Sequencing_Multiisolate', 'phs0004*'],
                'disease_abbr': ['BLCA', 'OR', 'BRCA'],
                'state': 'live'},
            'kwargs': {},
            'url': 'http://127.0.0.1:8983/solr/select/?'
                'q=study%3A(phs000178+OR+%2AOther_Sequencing_Multiisolate+OR+'
                'phs0004%2A)+disease_abbr%3A(BLCA+OR+OR+OR+BRCA)+state%3Alive&rows=1000000'
        }, {
            'query': {
                'study': '*Other_Sequencing_Multiisolate',
                'state': ['live'],
                'last_modified': '[NOW-1MONTH+TO+NOW]'},
            'kwargs': {},
            'url': 'http://127.0.0.1:8983/solr/select/?'
                'q=study%3A%2AOther_Sequencing_Multiisolate+'
                'last_modified%3A%5BNOW-1MONTH%2BTO%2BNOW%5D+state%3A(live)&rows=1000000'
        }, {
            'query': {
                'study': '*Other_Sequencing_Multiisolate',
                'state': ['live']},
            'kwargs': {'limit': 10, 'offset': 10, 'sort_by': '-state'},
            'url': 'http://127.0.0.1:8983/solr/select/?'
                'q=study%3A%2AOther_Sequencing_Multiisolate+'
                'state%3A(live)&start=10&rows=10&sort=state+desc'
        }
    ]


if __name__ == '__main__':
    unittest.main()


cghub_python_api/utils.py
import time

import urllib2


def urlopen(url, format='xml', max_attempts=5, sleep_time=1):
    """
    Wrapper for urllib2.urlopen.
    Retry getting answer from server if URLError raised (limited number of attempts).

    :param url: url
    :param format: 'xml' or 'json'
    :param max_attempts: maximum count of attempts to retrieve answer after URLError raised
    :param sleep_time: time between attempts, seconds
    """

    FORMATS = {
        'xml': 'text/xml',
        'json': 'application/json',
    }

    error_msg = ''
    for i in range(max_attempts):
        try:
            req = urllib2.Request(url, headers={
                    'Accept':  FORMATS.get(format, FORMATS['xml'])})
            return urllib2.urlopen(req)
        except urllib2.URLError as e:
            try:
                if e.getcode() == 400:
                    raise urllib2.URLError(u'%s: %s' % (e.msg, url))
                error_msg = e.msg
            except AttributeError:
                pass
            time.sleep(sleep_time)
    if error_msg:
        error_msg = ' (%s)' % error_msg
    raise urllib2.URLError(
            u'No response after %d attempts: %s%s' % (
                    max_attempts, url, error_msg))


cghub_python_api/api.py
import urllib2
import ijson

from xml.dom import pulldom
from xml.etree import ElementTree

from .utils import urlopen


DATE_ATTRIBUTES = ('last_modified', 'upload_date', 'published_date',)


class NonExistent(object):
    """
    Object returns None for not existent attributes
    instead of rising AttributeError.
    """

    exist = False

    def __getattr__(self, attr):
        if attr in self.__dict__:
            return getattr(self, attr)
        return None


class Result(object):
    """
    Wrapper for ElementTree object.
    Allows to access to tree nodes using attributes.

    Usage example:
    result = Result(tree)
    print result.analysis_id
    print result['analysis_id']
    print result['filename.0']
    print result['filename.1']

    If requested node was not found, will be returned NonExistent object.
    To check is requested node exists can be used 'exist' attribute.
    result.files.exist == True
    result.badattr.exist == False
    """

    API_SOLR = 'sorl'
    API_WSAPI = 'wsapi'

    def __init__(self, tree):
        """
        :param tree: ElementTree object
        """
        self.tree = tree
        if self.tree.find('analysis_id') is None:
            self.api = self.API_SOLR
        else:
            self.api = self.API_WSAPI

    def get_node(self, attr, index):
        if self.api == self.API_WSAPI:
            nodes = self.tree.findall('.//%s' % attr)
        else:
            nodes = self.tree.findall(".//*[@name='%s']" % attr)
            # open arrays
            if len(nodes) == 1 and nodes[0].tag == 'arr':
                nodes = nodes[0].findall('.//*')
        if len(nodes) < index + 1:
            return NonExistent()
        nodes[index].exist = True
        return nodes[index]

    def __getattr__(self, attr):
        if attr in self.__dict__:
            return getattr(self, attr)
        return self.get_node(attr, 0)

    def __getitem__(self, path):
        if '.' in path:
            try:
                attr, index = path.split('.')
                index = int(index)
            except ValueError:
                return NonExistent()
        else:
            attr = path
            index = 0
        return self.get_node(attr, index)


class BaseRequest(object):
    """
    CGHub API base class
    """

    FORMAT_XML = 'xml'
    FORMAT_JSON = 'json'

    def __init__(self,
            query, offset=0, limit=None, sort_by=None,
            server_url=None, uri=None, fields=None, format=FORMAT_XML):
        """
        :param query: a dict with query to send to the server
        :param offset: how many results should be skipped
        :param limit: how many records output should have
        :param sort_by: the attribute by which the results should be sorted (use '-' for reverse)
        :param server_url: server url where API works
        :param uri: uri that will be used to access data on API server
        :param fields: list of fields will be returned in result (used only in SOLRRequest)
        :param format: format of requested data ('xml' or 'json')
        """
        self.query = query
        self.server_url = server_url
        self.uri = uri
        self.offset = offset
        self.limit = limit
        self.sort_by = sort_by
        self.fields = fields
        self.format = format

    def get_source_file(self, url):
        """
        Can be overridden for testing purposes
        """
        return urlopen(url, format=self.format)

    def patch_input_data(self):
        """
        Here can be changed self.server_url, self.uri and other input data
        """
        pass

    def patch_result(self, result, result_xml):
        """
        Can be overriden to add custom fields to results, etc.
        """
        return result

    def call(self):
        """
        Makes a request to cghub server.
        Returns generator that returns Result objects.
        """
        raise NotImplementedError()

    def escape_query_value(self, key, value):
        val = urllib2.unquote(str(value)).replace('\-', '-')
        if key not in DATE_ATTRIBUTES:
            val = val.replace(' ', '+').replace('+OR+', ' OR ')
        return urllib2.quote(val)

    def build_query(self):
        """
        Builds query to access to cghub server.
        """
        raise NotImplementedError()


class WSAPIRequest(BaseRequest):
    """
    Allow to access to cghub data by query.

    Usage example:
    request = Request(query={
            'analysis_id': '017a4d4e-9f4b-4904-824e-060fde3ca223'})
    for result in request.call():
        print result.state
    """

    CGHUB_SERVER = 'https://192.35.223.223'
    CGHUB_ANALYSIS_ID_URI = '/cghub/metadata/analysisId'
    CGHUB_ANALYSIS_DETAIL_URI = '/cghub/metadata/analysisDetail'
    CGHUB_ANALYSIS_FULL_URI = '/cghub/metadata/analysisFull'
    FORMAT_XML = 'xml'
    FORMAT_JSON = 'json'

    def __init__(self,
                query, offset=0, limit=None, sort_by=None,
                server_url=CGHUB_SERVER, uri=CGHUB_ANALYSIS_DETAIL_URI,
                fields=None, format=FORMAT_XML):
        """
        :param query: a dict with query to send to the server
        :param offset: how many results should be skipped
        :param limit: how many records output should have
        :param sort_by: the attribute by which the results should be sorted (use '-' for reverse)
        :param server_url: server url where WSAPI works
        :param uri: uri that will be used to access data on WSAPI server
        :param fields: list of fields will be returned in result (used only in SOLRRequest)
        :param format: format of requested data from wsapi server ('xml' or 'json')
        """
        self.query = query
        self.server_url = server_url
        self.uri = uri
        self.offset = offset
        self.limit = limit
        self.sort_by = sort_by
        self.format = format

    def call(self):
        """
        Makes a request to cghub server.
        Returns generator that returns Result objects.
        """
        self.patch_input_data()
        query = self.build_query()
        url = '%s%s' % (self.server_url, self.uri)
        if query:
            url = '%s?%s' % (url, query)
        xml = self.get_source_file(url)
        if self.format == self.FORMAT_JSON:
            results = ijson.items(xml, 'result_set.results.item')
            for item in results:
                yield item
        else:
            # http://docs.python.org/dev/library/xml.dom.pulldom.html
            doc = pulldom.parse(xml)
            for event, node in doc:
                if event == pulldom.START_ELEMENT:
                    if node.tagName == 'Result':
                        doc.expandNode(node)
                        # convert to python object
                        # http://docs.python.org/2/library/xml.etree.elementtree.html
                        result_xml = node.toxml(encoding='utf-8')
                        tree = ElementTree.fromstring(result_xml)
                        result = Result(tree)
                        yield self.patch_result(result, result_xml)
                    elif node.tagName == 'Hits':
                        doc.expandNode(node)
                        tree = ElementTree.fromstring(node.toxml())
                        self.hits = int(tree.text)

    def build_query(self):
        """
        Builds query to access to cghub server.
        """
        parts = []
        for key, value in self.query.iteritems():
            if isinstance(value, list) or isinstance(value, tuple):
                value_str = '+OR+'.join([
                        self.escape_query_value(key, v) for v in value])
                value_str = '(%s)' % value_str
            else:
                value_str = self.escape_query_value(key, value)
            parts.append('='.join([key, value_str]))
        if self.offset:
            parts.append('='.join(['start', str(self.offset)]))
        if self.limit:
            parts.append('='.join(['rows', str(self.limit)]))
        if self.sort_by:
            if self.sort_by[0] == '-':
                parts.append('='.join([
                        'sort_by',
                        '%s:desc' % urllib2.quote(self.sort_by[1:])]))
            else:
                parts.append('='.join([
                        'sort_by',
                        '%s:asc' % urllib2.quote(self.sort_by)]))
        return '&'.join(parts)


class SOLRRequest(BaseRequest):
    """
    Allow to access to cghub data by query.

    Usage example:
    request = Request(query={
            'analysis_id': '017a4d4e-9f4b-4904-824e-060fde3ca223'})
    for result in request.call():
        print result.state
    """

    CGHUB_SERVER = 'http://127.0.0.1:8983'
    CGHUB_SEARCH_URI = '/solr/select/'
    FORMAT_XML = 'xml'
    FORMAT_JSON = 'json'

    def __init__(self,
                query, offset=0, limit=None, sort_by=None,
                server_url=CGHUB_SERVER, uri=CGHUB_SEARCH_URI,
                fields=None, format=FORMAT_XML):
        """
        :param query: a dict with query to send to the server
        :param offset: how many results should be skipped
        :param limit: how many records output should have
        :param sort_by: the attribute by which the results should be sorted (use '-' for reverse)
        :param server_url: server url where WSAPI works
        :param uri: uri that will be used to access data on WSAP
        :param format: format of requested data from solr server ('xml' or 'json')
        """
        self.query = query
        self.server_url = server_url
        self.uri = uri
        self.offset = offset
        self.limit = limit
        self.sort_by = sort_by
        self.fields = fields
        self.format = format

    def call(self):
        """
        Makes a request to cghub server.
        Returns generator that returns Result objects.
        """
        self.patch_input_data()
        query = self.build_query()
        url = '%s%s' % (self.server_url, self.uri)
        if query:
            url = '%s?%s' % (url, query)
        xml = self.get_source_file(url)
        if self.format == self.FORMAT_JSON:
            results = ijson.items(xml, 'response.docs.item')
            for item in results:
                yield item
        else:
            # http://docs.python.org/dev/library/xml.dom.pulldom.html
            doc = pulldom.parse(xml)
            for event, node in doc:
                if event == pulldom.START_ELEMENT:
                    if node.tagName == 'doc':
                        doc.expandNode(node)
                        # convert to python object
                        # http://docs.python.org/2/library/xml.etree.elementtree.html
                        result_xml = node.toxml(encoding='utf-8')
                        tree = ElementTree.fromstring(result_xml)
                        result = Result(tree)
                        yield self.patch_result(result, result_xml)
                    elif node.tagName == 'result':
                        self.hits = int(node.getAttribute('numFound'))

    def build_query(self):
        """
        Builds query to access to cghub server.
        """
        parts = []
        for key, value in self.query.iteritems():
            if isinstance(value, list) or isinstance(value, tuple):
                value_str = '+OR+'.join([
                        self.escape_query_value(key, v) for v in value])
                value_str = '(%s)' % value_str
            else:
                value_str = self.escape_query_value(key, value)
            parts.append(urllib2.quote(':').join([key, value_str]))
        if len(parts):
            parts = ['q=%s' % '+'.join(parts)]
        else:
            parts = ['q=*%s*' % urllib2.quote(':')]
        if self.offset:
            parts.append('='.join(['start', str(self.offset)]))
        if self.limit:
            parts.append('='.join(['rows', str(self.limit)]))
        else:
            parts.append('rows=1000000')
        if self.sort_by:
            if self.sort_by[0] == '-':
                parts.append('='.join([
                        'sort',
                        '%s+desc' % urllib2.quote(self.sort_by[1:])]))
            else:
                parts.append('='.join([
                        'sort',
                        '%s+asc' % urllib2.quote(self.sort_by)]))
        if self.fields:
            parts.append('='.join([
                        'fl',
                        ','.join(
                            [urllib2.quote(f) for f in self.fields])]))
        if self.format == self.FORMAT_JSON:
            parts.append('wt=json')
        return '&'.join(parts)
